heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -102,31 +119,39 @@ private void onExit() {
   public abstract void wakeUp();
 
   public void addTasksOnWakeup(Runnable task) {
-    tasksOnWakeup.add(task);
-    // We need to wake up the looper itself when we add a new task, otherwise, it is possible
-    // this task will never be executed due to the looper will never be wake up.
-    wakeUp();
+    if (!lockAddNewTasks) {
```

Comment:
QQ: Can tasks be added by other thread?

If they can, then there might be a racing condition.
the existing tasks are being cleared, and a new package is received and some tasks might be added. A lock flag is not enough and we have to sync up the two thread.